src/librustc_mir/build/scope.rs

```diff
@@ -324,20 +359,22 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
     /// drops onto the end of `block` that are needed.  This must
     /// match 1-to-1 with `push_scope`.
     pub fn pop_scope(&mut self,
-                     extent: CodeExtent,
+                     extent: (CodeExtent, SourceInfo),
                      mut block: BasicBlock)
                      -> BlockAnd<()> {
         debug!("pop_scope({:?}, {:?})", extent, block);
         // We need to have `cached_block`s available for all the drops, so we call diverge_cleanup
         // to make sure all the `cached_block`s are filled in.
-        self.diverge_cleanup();
+        self.diverge_cleanup(extent.1.span);
         let scope = self.scopes.pop().unwrap();
-        assert_eq!(scope.extent, extent);
+        assert_eq!(scope.extent, extent.0);
         unpack!(block = build_scope_drops(&mut self.cfg,
                                           &scope,
                                           &self.scopes,
                                           block,
                                           self.arg_count));
+
+        self.cfg.push_end_region(block, extent.1, scope.extent);
```

Comment:
my goal had been to continue lazily constructing the diverge path.

Its easy to generate EndRegion statements on their own (if somewhat wasteful) and then delete them afterwards when they are trivial.

I don't know whether it is as easy to do something similar for a diverge path that has been unnecessarily introduced. And I don't think the kind of example that arielb1 showed (that led me to stop trying to prefilter `EndRegion` in general) exists for diverge paths.